### PR TITLE
Expose auto-IO-throttle from Lucene's ConcurrentMergeScheduler

### LINKED ISSUE #9133: Expose ConcurrentMergeScheduler's auto IO throttle & stats
With https://issues.apache.org/jira/browse/LUCENE-6119 we're enabling Lucene's ConcurrentMergeScheduler to automatically set an adaptive IO throttle (writes) over time depending on the incoming merge load.

With this, an app that indexes at a slow rate but unluckily suddenly hits a large merge will see that merge aggressively throttled (not using much IO), but apps with intense indexing will see merges run at a higher IO rate or entirely unthrottled, so merges don't fall behind.

The only exposed setting is a boolean (default is true) to enable the adaptive IO throttle; it's a live setting, so I'll expose it as a dynamic setting in Elasticsearch.  I'll also remove the current hardwired throttling: it's too simplistic / dangerous.

With this change, merges also separately track how much time they were stopped (because too many smaller merges need to run) vs paused (due to IO throttle).  I'll cutover Elasticsearch's stats tracking to these...

### PR DESCRIPTION
**Pull request by mikemccand**
This removes all index/indices store level rate limiting (indices.store.throttle.type, indices.store.throttle.max_bytes_per_sec, index.store.throttle.type, index.store.throttle.max_bytes_per_sec) and cuts over to Lucene's auto-throttle boolean (default: on) in ConcurrentMergeScheduler added in https://issues.apache.org/jira/browse/LUCENE-6119

I added a new live boolean setting (index.merge.scheduler.auto_throttle), default is on (like Lucene).

This also removes throttle_time from store stats, and adds total_throttled_time, total_stopped_time (total time when large merges were stopped so smaller merges could finish) and total_throttled_bytes_per_sec (the current bytes/sec throttle) stats to merge stats.  Merge logging also shows the stopped/throttle time per merge.

Recovery/snapshot/restore still have their rate limiters and still default to 20 MB/sec.

This should also fix the slowdowns / index size increase from http://benchmarks.elasticsearch.org and I think improve indexing performance at the defaults since auto throttling should prevent the merge backlog and index throttling.

Closes #9133

### CODE REVIEW

File: src/main/java/org/elasticsearch/index/merge/MergeStats.java [hunk 1]
```diff
@@ -176,5 +235,10 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeVLong(current);
         out.writeVLong(currentNumDocs);
         out.writeVLong(currentSizeInBytes);
+        if (out.getVersion().onOrAfter(Version.V_2_0_0)) {
```
Comment (martijnvg) [inline src/main/java/org/elasticsearch/index/merge/MergeStats.java]: Upgrading from 1.x to 2.x will require a full cluster restart, so a 2.x node will never chat with a < 2.x node. Therefor the version checks are redundant and can be removed. (assuming that this change only goes into master)

**Comment by rmuir** [comment]
looks good

**Comment by s1monw** [comment]
LGTM too. I wonder if somebody had disabled the merge throtteling in 1.x and they upgrade should they also get this disable. I am leaning towards keeping it `true` / `enable` anyway I just wanted to throw it out there

File: src/main/java/org/elasticsearch/index/merge/MergeStats.java (hunk 1 above)
Comment (mikemccand) [inline src/main/java/org/elasticsearch/index/merge/MergeStats.java]: Aha!  OK I'll remove the version check...

**Comment by mikemccand** [comment]
> I wonder if somebody had disabled the merge throtteling in 1.x and they upgrade should they also get this disable. I am leaning towards keeping it true / enable anyway I just wanted to throw it out there

I think it's OK to keep the default (true) in this case: auto-throttle should still move to un-throttled if merges can't keep up.  For http://benchmarks.elasticsearch.org I will run the fast/fastupdates with auto-throttle enabled.

**Comment by s1monw** [comment]
ok fair enough

**Comment by bleskes** [comment]
@mikemccand I was wondering how this change works if there are multiple shards on a node - previously the rate limiting was a node level settings, making sure that we do not overload the node even it has multiple shards on it merging together. If I understand correctly it is no possible that a heavy loaded index will consume more and more node resources before starting to push back on indexing, potentially starving other shards that has nothing to do with this index. I might be missing something here.

Second think I wondering what we should do when someone turns off auto_throttling - to me it seems we should fall back to non-auto throttling (like we had) as opposed to no throttling at all? if that's what we want - might as well call it `index.merge.scheduler.throttle` . Again - I might be missing something here.

**Comment by mikemccand** [comment]
> I was wondering how this change works if there are multiple shards on a node

Yes, this is a difference: the auto-throttle is only within a single
shard vs current rate limiter which can set aggregate for the whole
node.

But I think this is dangerous: each shard really needs its merges to
keep up with the indexing rate against that shard, so it remains
healthy.  So I think each shard should look out for itself / increase
its throttle to keep up w/ indexing it's seeing.

> Second think I wondering what we should do when someone turns off auto_throttling - to me it seems we should fall back to non-auto throttling

I think hardwired merge throttling is too simplistic: who can properly
predict what merge MB/sec is required?  So net/net I don't think we
should keep this feature long term.

But short term: I agree we should keep it.  It's safer to have both
options, e.g. in case there are horrible bugs in auto-throttling, apps
can easily fallback.  I'll rework the PR to keep both...

**Comment by mikemccand** [comment]
OK I put back the fixed rate limiting, but left it off by default in favor of auto-throttle.  I think it's ready.

**Comment by jpountz** [comment]
Since the intention is to move away from fixed rate limiting and using auto throttling instead, maybe documentation about fixed rate limiting should be removed in order not to encourage using it?

Otherwise LGTM!

**Comment by mikemccand** [comment]
That makes sense @jpountz, I'll remove the docs.

**Pull request merged**